src/FastMathFunctions.cpp

```diff
@@ -0,0 +1,872 @@
+#include "FastMathFunctions.h"
+
+#include "ApproximationTables.h"
+#include "CSE.h"
+#include "IRMutator.h"
+#include "IROperator.h"
+#include "IRPrinter.h"
+
+namespace Halide {
+namespace Internal {
+
+namespace {
+
+Expr constant(Type t, double value) {
+    if (t == Float(64)) {
+        return Expr(value);
+    }
+    if (t == Float(32)) {
+        return Expr(float(value));
+    }
+    internal_error << "Constants only for double or float.";
+    return 0;
+}
+
+}  // namespace
+
+namespace ApproxImpl {
+
+constexpr double PI = 3.14159265358979323846;
+constexpr double ONE_OVER_PI = 1.0 / PI;
+constexpr double TWO_OVER_PI = 2.0 / PI;
+constexpr double PI_OVER_TWO = PI / 2;
+
+Expr eval_poly(const std::vector<double> &coefs, const Expr &x) {
+    /*
+     * The general scheme looks like this:
+     *
+     * R = a0 + x * a1 + x^2 * a2 + x^3 * a3
```

Comment:
This is called Horner's method. main evaluates polynomials a little differently (see evaluate_polynomial in IROperator.cpp). We found that method to be faster, despite doing an additional multiply, because it has better instruction latencies.

Reply:
I adopted the function from the main-branch, but I cannot reproduce this method to be faster. I wouldn't be surprised if they both compile to the same thing. Either by Halide simplifications or LLVM optimizations. Tried this on both host (x64) and host-vulkan.

Reply:
Comparing assembly side-by-side gives indeed the exact same assembly for me.